File: python-sdk/ag_ui/encoder/encoder.py
"""
This module contains the EventEncoder class
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ag_ui.core.events import BaseEvent


AGUI_MEDIA_TYPE = "application/vnd.ag-ui.event+proto"

class EventEncoder:
    """
    Encodes Agent User Interaction events.
    """
    def __init__(self, accept: str | None = None) -> None:
        """
        Initializes the EventEncoder.
        """
        self.accept = accept

    def get_content_type(self) -> str:
        """
        Returns the content type of the encoder.
        """
        return "text/event-stream"

    def encode(self, event: BaseEvent) -> str:
        """
        Encodes an event.
        """
        return self._encode_sse(event)

    def _encode_sse(self, event: BaseEvent) -> str:
        """
        Encodes an event into an SSE string.
        """
        return f"data: {event.model_dump_json(by_alias=True, exclude_none=True)}\n\n"

File: sdks/nim/src/ag_ui_nim/encoder/encoder.nim
import std/[json, strformat]
import ../core/events

const AGUI_MEDIA_TYPE* = "application/vnd.ag-ui.event+proto"

type
  EventEncoder* = object
    acceptsProtobuf: bool

proc newEventEncoder*(accept: string = ""): EventEncoder =
  EventEncoder(acceptsProtobuf: false)

proc getContentType*(encoder: EventEncoder): string =
  if encoder.acceptsProtobuf:
    return AGUI_MEDIA_TYPE
  else:
    return "text/event-stream"

proc encodeSSE*[T](encoder: EventEncoder, event: T): string =
  let jsonStr = $event.toJson()
  result = fmt"data: {jsonStr}" & "\n\n"

proc encode*(encoder: EventEncoder, event: Event): string =
  encoder.encodeSSE(event)

proc encode*[T](encoder: EventEncoder, event: T): string =
  encoder.encodeSSE(event)

export newEventEncoder, getContentType, encode, encodeSSE